acvm-repo/acvm/src/pwg/mod.rs

```diff
@@ -258,13 +259,35 @@ impl<'backend, B: BlackBoxFunctionSolver> ACVM<'backend, B> {
                 solver.solve_memory_op(op, &mut self.witness_map, predicate)
             }
             Opcode::Brillig(brillig) => {
-                match BrilligSolver::solve(
-                    &mut self.witness_map,
-                    brillig,
-                    self.backend,
-                    self.instruction_pointer,
-                ) {
-                    Ok(Some(foreign_call)) => return self.wait_for_foreign_call(foreign_call),
+                let witness = &mut self.witness_map;
+                // get the active Brillig solver, or try to build one if necessary
+                // (Brillig execution maybe bypassed by constraints)
+                let maybe_solver = match self.brillig_solver.as_mut() {
+                    Some(solver) => Ok(Some(solver)),
+                    None => BrilligSolver::build_or_skip(
+                        witness,
+                        brillig,
+                        self.backend,
+                        self.instruction_pointer,
+                    )
+                    .map(|optional_solver| {
+                        optional_solver.map(|solver| self.brillig_solver.insert(solver))
+                    }),
+                };
+                match maybe_solver {
+                    Ok(Some(solver)) => match solver.solve() {
+                        Ok(BrilligSolverStatus::ForeignCallWait(foreign_call)) => {
+                            return self.wait_for_foreign_call(foreign_call);
+                        }
+                        Ok(BrilligSolverStatus::InProgress) => {
+                            unreachable!("Brillig solver still in progress")
+                        }
+                        Ok(BrilligSolverStatus::Finished) => {
+                            // clear active Brillig solver and write execution outputs
+                            self.brillig_solver.take().unwrap().finalize(witness, brillig)
+                        }
+                        res => res.map(|_| ()),
+                    },
                     res => res.map(|_| ()),
                 }
```

Comment:
For example, compare this with pseudocode

```rust
if BrilligSolver::should_skip(witness, brillig)? {
  BrilligSolver::zero_out_brillig_outputs(witness, brillig)
} else {
  let solver = self.brillig_solver.take().unwrap_or_else(|| {
      BrilligSolver::new(witness, brillig, self.backend, self.instruction_pointer)
  });
  match solver.solve()? {
      BrilligSolverStatus::ForeignCallWait(foreign_call) => {
          // Save solver state to resume execution after foreign call.
          self.brillig_solver.insert(solver);
          return self.wait_for_foreign_call(foreign_call);
      }
      BrilligSolverStatus::InProgress => {
          unreachable!("Brillig solver still in progress")
      }
      BrilligSolverStatus::Finished => {
          // Write execution outputs to witness
          solver.finalize(witness, brillig)
      }
  }
}
```

We'll need to turn this big `let resolution = match` statement into a helper function so that we can use `?` syntax, but it's much easier to see what this code is doing imo.

Reply:
I like this! But there is one problem for which I haven't found a good idiomatic solution: the early return when waiting for foreign calls, which cannot go in the helper. I'm still familiarizing myself with Rust and don't know of a good way to handle this, short of introducing yet another return enum type (which I'd like to avoid). Maybe there is something in the stdlib?

Reply:
Ah yeah, I hadn't thought of that. Maybe for the short term we could just move the brillig arm of this match into a helper function and return a `Result<Option<ForeignCallWaitInfo>, OpcodeResolutionError>`. We can then perform the early return based on whether the option is `Some` or `None`.

```rust
// New match arm
match self.solve_brillig_opcode(brillig) {
    Ok(Some(foreign_call)) => {
        return self.wait_for_foreign_call(foreign_call);
     }
     result => result.map(|_| ()),
}

// helper function
fn solve_brillig_opcode(
    &mut self,
    brillig: &Brillig,
) -> Result<Option<ForeignCallWaitInfo>, OpcodeResolutionError> {
    let witness = &mut self.witness_map;
    if BrilligSolver::should_skip(witness, brillig)? {
        BrilligSolver::zero_out_brillig_outputs(witness, brillig).map(|_| None)
    } else {
        let solver = self.brillig_solver.take().unwrap_or_else(|| {
            BrilligSolver::new(witness, brillig, self.backend, self.instruction_pointer)
        });
        match solver.solve()? {
            BrilligSolverStatus::ForeignCallWait(foreign_call) => {
                // Save solver state to resume execution after foreign call.
               self.brillig_solver.insert(solver);
                Ok(Some(foreign_call))
            }
            BrilligSolverStatus::InProgress => {
                unreachable!("Brillig solver still in progress")
            }
            BrilligSolverStatus::Finished => {
                // Write execution outputs to witness
                solver.finalize(witness, brillig).map(|_| None)
            }
        }
    }
}
```

Reply:
I steered the code in the direction you proposed, but I found a couple other problems:
- It's not possible to use `unwrap_or_else` because `BrilligSolver::new` may return an error when reading the inputs.
- I found no clean way to send `brillig` as an argument, because that's a reference inside `self`, which needs to be mutably borrowed for the call itself.

I'm not 100% happy with the resulting code, but I think it's better than before. In any case, it will probably mutate again when we add support for step-debugging through Brillig blocks.

Reply:
Yeah, thanks for making these changes. I'm happy with the current state of this code, it's not perfect but given the constraints I think it's as readable as we can get for now. Let's leave it as is, as you say, we're going to be modifying it again in future.

> I found no clean way to send brillig as an argument, because that's a reference inside self, which needs to be mutably borrowed for the call itself.

Yeah, it's a little bit of a pain. If we passed in references to individual fields from `self` we could get around this but  that's just another tradeoff.